crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl

```diff
@@ -0,0 +1,168 @@
+    let frag_coords = vec2<i32>(in.uv * texture_size);
+
+#ifdef MULTISAMPLED
+    let base_color = textureLoad(screen_texture, frag_coords, i32(sample_index));
+#else
+    let base_color = textureSample(screen_texture, texture_sampler, in.uv);
+#endif
+
+    let shutter_angle = settings.shutter_angle;
+
+#ifdef MULTISAMPLED
+    let this_motion_vector = textureLoad(motion_vectors, frag_coords, i32(sample_index)).rg;
+#else
+    let this_motion_vector = textureSample(motion_vectors, texture_sampler, in.uv).rg;
+#endif
+
+#ifdef NO_DEPTH_TEXTURE_SUPPORT
+    let this_depth = 0.0;
+    let depth_supported = false;
+#else
+    let depth_supported = true;
+#ifdef MULTISAMPLED
+    let this_depth = textureLoad(depth, frag_coords, i32(sample_index));
+#else
+    let this_depth = textureSample(depth, texture_sampler, in.uv);
+#endif
+#endif
+    
+    // The exposure vector is the distance that this fragment moved while the camera shutter was
+    // open. This is the motion vector (total distance traveled) multiplied by the shutter angle (a
+    // fraction). In film, the shutter angle is commonly 0.5 or "180 degrees" (out of 360 total).
+    // This means that for a frame time of 20ms, the shutter is only open for 10ms.
+    //
+    // Using a shutter angle larger than 1.0 is non-physical, objects would need to move further
+    // than they physically travelled during a frame, which is not possible. Note: we allow values
+    // larger than 1.0 because it may be desired for artistic reasons.
+    let exposure_vector = shutter_angle * this_motion_vector;
+
+    var accumulator: vec4<f32>;
+    var weight_total = 0.0;
+    let n_samples = i32(settings.samples);
+    let noise = hash_noise(frag_coords, globals.frame_count); // 0 to 1
+       
+    for (var i = -n_samples; i < n_samples; i++) {
+        // The current sample step vector, from in.uv
+        let step_vector = 0.5 * exposure_vector * (f32(i) + noise) / f32(n_samples);
+        var sample_uv = in.uv + step_vector;
+        let sample_coords = vec2<i32>(sample_uv * texture_size);
+
+    #ifdef MULTISAMPLED
+        let sample_color = textureLoad(screen_texture, sample_coords, i32(sample_index));
+    #else
+        let sample_color = textureSample(screen_texture, texture_sampler, sample_uv);
+    #endif
+    #ifdef MULTISAMPLED
+        let sample_motion = textureLoad(motion_vectors, sample_coords, i32(sample_index)).rg;
+    #else
+        let sample_motion = textureSample(motion_vectors, texture_sampler, sample_uv).rg;
+    #endif
+    #ifdef NO_DEPTH_TEXTURE_SUPPORT
+        let sample_depth = 0.0;
+    #else
+    #ifdef MULTISAMPLED
+        let sample_depth = textureLoad(depth, sample_coords, i32(sample_index));
+    #else
+        let sample_depth = textureSample(depth, texture_sampler, sample_uv);
+    #endif
+    #endif
+
+        var weight = 1.0;
+        let is_sample_in_fg = !(depth_supported && sample_depth < this_depth && sample_depth > 0.0);
+        if is_sample_in_fg {
```

Comment:
Rather than using an if statement, you can also turn it into a binary 0/1 weight that you multiply the result by to avoid branching. Depends on how coherent you think threadgroups will be.

Reply:
I don't think that can make a difference either way here, there is no else clause, so no matter what, worst case scenario is that the thread has to compute what is inside the block, which is equivalent to dissolving the block, computing it every time, and multiplying by a weight.

Reply:
Makes sense